File: crates/nodebox-ndbx/src/parser.rs
//! NDBX file parser.

use std::fs;
use std::path::Path;

use quick_xml::events::{BytesStart, Event};
use quick_xml::Reader;

use nodebox_core::node::{Connection, MenuItem, Node, NodeLibrary, Port, PortRange, PortType, Widget};
use nodebox_core::geometry::Point;
use nodebox_core::Value;

use crate::error::{NdbxError, Result};

/// Parses an NDBX file from the given path.
pub fn parse_file(path: impl AsRef<Path>) -> Result<NodeLibrary> {
    let content = fs::read_to_string(path)?;
    parse(&content)
}

/// Parses NDBX content from a string.
pub fn parse(content: &str) -> Result<NodeLibrary> {
    let mut reader = Reader::from_str(content);
    reader.trim_text(true);

    let mut library = NodeLibrary::default();
    let mut buf = Vec::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(ref e)) => {
                if e.name().as_ref() == b"ndbx" {
                    parse_ndbx_element(&mut reader, e, &mut library)?;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(e.into()),
            _ => {}
        }
        buf.clear();
    }

    Ok(library)
}

fn parse_ndbx_element(
    reader: &mut Reader<&[u8]>,
    start: &BytesStart,
    library: &mut NodeLibrary,
) -> Result<()> {
    // Parse ndbx attributes
    for attr in start.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let value = std::str::from_utf8(&attr.value)?;

        match key {
            "formatVersion" => {
                // Handle both integer and float format versions (older files had "0.9" etc.)
                library.format_version = if let Ok(v) = value.parse::<u32>() {
                    v
                } else if let Ok(v) = value.parse::<f64>() {
                    // Convert float versions like 0.9 to 0
                    v.floor() as u32
                } else {
                    return Err(NdbxError::parse_value(value, "u32", "invalid format version"));
                };
            }
            "uuid" => {
                library.uuid = Some(value.to_string());
            }
            "type" => {
                // Store type in properties for now
                library.properties.insert("type".to_string(), value.to_string());
            }
            _ => {}
        }
    }

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(ref e)) => {
                let name = e.name();
                match name.as_ref() {
                    b"property" => {
                        parse_property(e, library)?;
                    }
                    b"link" => {
                        parse_link(e, library)?;
                    }
                    b"node" => {
                        let node = parse_node(reader, e)?;
                        library.root = node;
                    }
                    _ => {}
                }
            }
            Ok(Event::Empty(ref e)) => {
                let name = e.name();
                match name.as_ref() {
                    b"property" => {
                        parse_property(e, library)?;
                    }
                    b"link" => {
                        parse_link(e, library)?;
                    }
                    _ => {}
                }
            }
            Ok(Event::End(ref e)) => {
                if e.name().as_ref() == b"ndbx" {
                    break;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(e.into()),
            _ => {}
        }
        buf.clear();
    }

    Ok(())
}

fn parse_property(e: &BytesStart, library: &mut NodeLibrary) -> Result<()> {
    let mut name = String::new();
    let mut value = String::new();

    for attr in e.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let val = std::str::from_utf8(&attr.value)?;

        match key {
            "name" => name = val.to_string(),
            "value" => value = val.to_string(),
            _ => {}
        }
    }

    if !name.is_empty() {
        library.properties.insert(name, value);
    }

    Ok(())
}

fn parse_link(e: &BytesStart, library: &mut NodeLibrary) -> Result<()> {
    let mut href = String::new();
    let mut rel = String::new();

    for attr in e.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let val = std::str::from_utf8(&attr.value)?;

        match key {
            "href" => href = val.to_string(),
            "rel" => rel = val.to_string(),
            _ => {}
        }
    }

    // Store links as properties with prefix
    if !href.is_empty() {
        let key = format!("link.{}.{}", rel, href.replace(['/', ':'], "_"));
        library.properties.insert(key, href);
    }

    Ok(())
}

fn parse_node(reader: &mut Reader<&[u8]>, start: &BytesStart) -> Result<Node> {
    let mut node = parse_node_attributes(start)?;

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(ref e)) => {
                let name = e.name();
                match name.as_ref() {
                    b"node" => {
                        let child = parse_node(reader, e)?;
                        node.children.push(child);
                    }
                    b"port" => {
                        let port = parse_port(reader, e)?;
                        node.inputs.push(port);
                    }
                    _ => {
                        // Skip unknown elements
                        skip_element(reader, &name)?;
                    }
                }
            }
            Ok(Event::Empty(ref e)) => {
                let name = e.name();
                match name.as_ref() {
                    b"node" => {
                        let child = parse_node_attributes(e)?;
                        node.children.push(child);
                    }
                    b"port" => {
                        let port = parse_port_attributes(e)?;
                        node.inputs.push(port);
                    }
                    b"conn" => {
                        if let Some(conn) = parse_connection(e)? {
                            node.connections.push(conn);
                        }
                    }
                    _ => {}
                }
            }
            Ok(Event::End(ref e)) => {
                if e.name().as_ref() == b"node" {
                    break;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(e.into()),
            _ => {}
        }
        buf.clear();
    }

    Ok(node)
}

fn parse_node_attributes(e: &BytesStart) -> Result<Node> {
    let mut node = Node::default();

    for attr in e.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let value = std::str::from_utf8(&attr.value)?;

        match key {
            "name" => node.name = value.to_string(),
            "prototype" => node.prototype = Some(value.to_string()),
            "function" => node.function = Some(value.to_string()),
            "category" => node.category = value.to_string(),
            "description" => node.description = Some(value.to_string()),
            "position" => {
                node.position = parse_point(value)?;
            }
            "outputType" => {
                node.output_type = parse_port_type(value);
            }
            "outputRange" => {
                node.output_range = parse_port_range(value);
            }
            "renderedChild" => {
                node.rendered_child = Some(value.to_string());
            }
            "handle" => {
                node.handle = Some(value.to_string());
            }
            "alwaysRendered" => {
                node.always_rendered = value == "true";
            }
            "image" => {
                // Store image in a way that can be retrieved later
                // For now we'll skip it as it's UI-related
            }
            _ => {}
        }
    }

    // Nodes with children are networks
    // This will be set properly if we encounter child nodes
    if node.prototype.as_deref() == Some("core.network") {
        node.is_network = true;
    }

    Ok(node)
}

fn parse_port(reader: &mut Reader<&[u8]>, start: &BytesStart) -> Result<Port> {
    let mut port = parse_port_attributes(start)?;

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Empty(ref e)) => {
                if e.name().as_ref() == b"menu" {
                    if let Some(item) = parse_menu_item(e)? {
                        port.menu_items.push(item);
                    }
                }
            }
            Ok(Event::End(ref e)) => {
                if e.name().as_ref() == b"port" {
                    break;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(e.into()),
            _ => {}
        }
        buf.clear();
    }

    Ok(port)
}

fn parse_port_attributes(e: &BytesStart) -> Result<Port> {
    let mut name = String::new();
    let mut port_type = PortType::Float;
    let mut range = PortRange::Value;
    let mut widget = None;
    let mut value_str = String::new();
    let mut min = None;
    let mut max = None;
    let mut label = None;
    let mut description = None;

    for attr in e.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let val = std::str::from_utf8(&attr.value)?;

        match key {
            "name" => name = val.to_string(),
            "type" => port_type = parse_port_type(val),
            "range" => range = parse_port_range(val),
            "widget" => widget = Some(parse_widget(val)),
            "value" => value_str = val.to_string(),
            "min" => min = val.parse().ok(),
            "max" => max = val.parse().ok(),
            "label" => label = Some(val.to_string()),
            "description" => description = Some(val.to_string()),
            _ => {}
        }
    }

    // Parse the value based on type
    let value = parse_value(&value_str, &port_type);

    let mut port = Port {
        name,
        port_type,
        range,
        widget: widget.unwrap_or(Widget::None),
        value,
        min,
        max,
        label,
        description,
        menu_items: Vec::new(),
    };

    // If widget wasn't specified, infer from type
    if widget.is_none() {
        port.widget = Widget::default_for_type(&port.port_type);
    }

    Ok(port)
}

fn parse_menu_item(e: &BytesStart) -> Result<Option<MenuItem>> {
    let mut key = String::new();
    let mut label = String::new();

    for attr in e.attributes().flatten() {
        let k = std::str::from_utf8(attr.key.as_ref())?;
        let v = std::str::from_utf8(&attr.value)?;

        match k {
            "key" => key = v.to_string(),
            "label" => label = v.to_string(),
            _ => {}
        }
    }

    if key.is_empty() {
        return Ok(None);
    }

    Ok(Some(MenuItem { key, label }))
}

fn parse_connection(e: &BytesStart) -> Result<Option<Connection>> {
    let mut input = String::new();
    let mut output = String::new();

    for attr in e.attributes().flatten() {
        let key = std::str::from_utf8(attr.key.as_ref())?;
        let value = std::str::from_utf8(&attr.value)?;

        match key {
            "input" => input = value.to_string(),
            "output" => output = value.to_string(),
            _ => {}
        }
    }

    // Parse input="node.port" format
    if let Some((input_node, input_port)) = input.split_once('.') {
        Ok(Some(Connection::new(
            &output,
            input_node,
            input_port,
        )))
    } else {
        Ok(None)
    }
}

fn parse_point(s: &str) -> Result<Point> {
    let parts: Vec<&str> = s.split(',').collect();
    if parts.len() == 2 {
        let x = parts[0].trim().parse().unwrap_or(0.0);
        let y = parts[1].trim().parse().unwrap_or(0.0);
        Ok(Point::new(x, y))
    } else {
        Ok(Point::ZERO)
    }
}

fn parse_port_type(s: &str) -> PortType {
    match s.to_lowercase().as_str() {
        "int" => PortType::Int,
        "float" => PortType::Float,
        "string" => PortType::String,
        "boolean" | "bool" => PortType::Boolean,
        "point" => PortType::Point,
        "color" => PortType::Color,
        "geometry" => PortType::Geometry,
        "list" => PortType::List,
        "context" => PortType::Context,
        "state" => PortType::State,
        _ => PortType::Custom(s.to_string()),
    }
}

fn parse_port_range(s: &str) -> PortRange {
    match s.to_lowercase().as_str() {
        "value" => PortRange::Value,
        "list" => PortRange::List,
        _ => PortRange::Value,
    }
}

fn parse_widget(s: &str) -> Widget {
    match s.to_lowercase().as_str() {
        "int" => Widget::Int,
        "float" => Widget::Float,
        "angle" => Widget::Angle,
        "toggle" => Widget::Toggle,
        "color" => Widget::Color,
        "menu" => Widget::Menu,
        "file" => Widget::File,
        "font" => Widget::Font,
        "text" => Widget::Text,
        "code" => Widget::Text,
        "point" => Widget::Point,
        "data" => Widget::Data,
        "image" => Widget::Image,
        "seed" => Widget::Seed,
        "none" => Widget::None,
        _ => Widget::None,
    }
}

fn parse_value(s: &str, port_type: &PortType) -> Value {
    if s.is_empty() {
        return Value::Null;
    }

    match port_type {
        PortType::Int => s.parse().map(Value::Int).unwrap_or(Value::Null),
        PortType::Float => s.parse().map(Value::Float).unwrap_or(Value::Null),
        PortType::String => Value::String(s.to_string()),
        PortType::Boolean => Value::Boolean(s == "true"),
        PortType::Point => {
            parse_point(s)
                .map(Value::Point)
                .unwrap_or(Value::Null)
        }
        PortType::Color => {
            nodebox_core::Color::from_hex(s)
                .map(Value::Color)
                .unwrap_or(Value::Null)
        }
        _ => Value::String(s.to_string()),
    }
}

fn skip_element(reader: &mut Reader<&[u8]>, name: &quick_xml::name::QName) -> Result<()> {
    let mut depth = 1;
    let mut buf = Vec::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(_)) => depth += 1,
            Ok(Event::End(ref e)) => {
                depth -= 1;
                if depth == 0 && e.name() == *name {
                    break;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(e.into()),
            _ => {}
        }
        buf.clear();
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_simple_ndbx() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test-uuid">
    <property name="canvasWidth" value="800"/>
    <property name="canvasHeight" value="600"/>
    <node name="root" prototype="core.network" renderedChild="rect1">
        <node name="rect1" position="1.00,2.00" prototype="corevector.rect">
            <port name="width" type="float" value="100.0"/>
            <port name="height" type="float" value="50.0"/>
        </node>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();

        assert_eq!(library.format_version, 17);
        assert_eq!(library.uuid, Some("test-uuid".to_string()));
        assert_eq!(library.canvas_width(), 800.0);
        assert_eq!(library.canvas_height(), 600.0);
        assert_eq!(library.root.name, "root");
        assert_eq!(library.root.rendered_child, Some("rect1".to_string()));
        assert_eq!(library.root.children.len(), 1);

        let rect = &library.root.children[0];
        assert_eq!(rect.name, "rect1");
        assert_eq!(rect.position, Point::new(1.0, 2.0));
        assert_eq!(rect.prototype, Some("corevector.rect".to_string()));
        assert_eq!(rect.inputs.len(), 2);
    }

    #[test]
    fn test_parse_connections() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network" renderedChild="colorize1">
        <node name="rect1" prototype="corevector.rect"/>
        <node name="colorize1" prototype="corevector.colorize"/>
        <conn input="colorize1.shape" output="rect1"/>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();
        assert_eq!(library.root.connections.len(), 1);

        let conn = &library.root.connections[0];
        assert_eq!(conn.output_node, "rect1");
        assert_eq!(conn.input_node, "colorize1");
        assert_eq!(conn.input_port, "shape");
    }

    #[test]
    fn test_parse_port_with_menu() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network">
        <node name="align1" prototype="corevector.align">
            <port name="halign" type="string" value="center" widget="menu">
                <menu key="left" label="Left"/>
                <menu key="center" label="Center"/>
                <menu key="right" label="Right"/>
            </port>
        </node>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();
        let align = &library.root.children[0];
        assert_eq!(align.inputs.len(), 1);

        let port = &align.inputs[0];
        assert_eq!(port.name, "halign");
        assert_eq!(port.widget, Widget::Menu);
        assert_eq!(port.menu_items.len(), 3);
        assert_eq!(port.menu_items[0].key, "left");
        assert_eq!(port.menu_items[1].label, "Center");
    }

    #[test]
    fn test_parse_color_value() {
        let content = r##"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network">
        <node name="colorize1" prototype="corevector.colorize">
            <port name="fill" type="color" value="#ff0000ff"/>
        </node>
    </node>
</ndbx>"##;

        let library = parse(content).unwrap();
        let colorize = &library.root.children[0];
        let fill = &colorize.inputs[0];

        if let Value::Color(c) = &fill.value {
            assert_eq!(c.r, 1.0);
            assert_eq!(c.g, 0.0);
            assert_eq!(c.b, 0.0);
            assert_eq!(c.a, 1.0);
        } else {
            panic!("Expected color value");
        }
    }

    #[test]
    fn test_parse_point_value() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network">
        <node name="rect1" prototype="corevector.rect">
            <port name="position" type="point" value="10.50,-20.25"/>
        </node>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();
        let rect = &library.root.children[0];
        let pos = &rect.inputs[0];

        if let Value::Point(p) = &pos.value {
            assert_eq!(p.x, 10.5);
            assert_eq!(p.y, -20.25);
        } else {
            panic!("Expected point value");
        }
    }

    #[test]
    fn test_parse_boolean_value() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network">
        <node name="connect1" prototype="corevector.connect">
            <port name="closed" type="boolean" value="true"/>
        </node>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();
        let connect = &library.root.children[0];
        let closed = &connect.inputs[0];

        assert_eq!(closed.value, Value::Boolean(true));
    }

    #[test]
    fn test_parse_empty_ndbx() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="21" type="file" uuid="empty">
    <node name="root" prototype="core.network"/>
</ndbx>"#;

        let library = parse(content).unwrap();
        assert_eq!(library.format_version, 21);
        assert_eq!(library.root.name, "root");
        assert!(library.root.children.is_empty());
    }

    #[test]
    fn test_parse_nested_networks() {
        let content = r#"<?xml version="1.0" encoding="UTF-8"?>
<ndbx formatVersion="17" type="file" uuid="test">
    <node name="root" prototype="core.network">
        <node name="subnet1" prototype="core.network">
            <node name="inner1" prototype="corevector.rect"/>
            <node name="inner2" prototype="corevector.ellipse"/>
        </node>
    </node>
</ndbx>"#;

        let library = parse(content).unwrap();
        assert_eq!(library.root.children.len(), 1);

        let subnet = &library.root.children[0];
        assert_eq!(subnet.name, "subnet1");
        assert!(subnet.is_network);
        assert_eq!(subnet.children.len(), 2);
        assert_eq!(subnet.children[0].name, "inner1");
        assert_eq!(subnet.children[1].name, "inner2");
    }
}

File: packages/nodebox-core/src/ndbx/parser.ts
import type { Node, Connection } from '../node/node.js';
import type { Port, PortType, Widget, PortRange, MenuItem } from '../node/port.js';
import type { Value } from '../node/value.js';
import type { NodeLibrary } from '../node/library.js';
import { CURRENT_FORMAT_VERSION, upgradeNdbx } from './upgrades.js';

export type LibraryLoader = (name: string) => NodeLibrary | undefined;

// Cache for loaded prototype libraries
const libraryCache = new Map<string, NodeLibrary>();

export function parseNdbx(xml: string, libraryLoader?: LibraryLoader): NodeLibrary {
  const doc = parseXml(xml);
  const root = doc.documentElement;

  if (root.tagName !== 'ndbx') {
    throw new Error(`Expected root element 'ndbx', got '${root.tagName}'`);
  }

  const formatVersion = parseInt(root.getAttribute('formatVersion') ?? String(CURRENT_FORMAT_VERSION), 10);
  const uuid = root.getAttribute('uuid') ?? '';

  // Upgrade if needed
  let xmlToUse = xml;
  if (formatVersion < CURRENT_FORMAT_VERSION) {
    const result = upgradeNdbx(xml, formatVersion);
    xmlToUse = result.xml;
    // Re-parse upgraded XML
    const upgradedDoc = parseXml(xmlToUse);
    return parseNdbxDocument(upgradedDoc, uuid, libraryLoader);
  }

  return parseNdbxDocument(doc, uuid, libraryLoader);
}

function parseNdbxDocument(doc: Document, uuid: string, libraryLoader?: LibraryLoader): NodeLibrary {
  const root = doc.documentElement;
  const properties: Record<string, string> = {};
  const functionLinks: string[] = [];

  // Parse properties and links
  for (let i = 0; i < root.childNodes.length; i++) {
    const child = root.childNodes[i];
    if (child.nodeType !== 1) continue; // Element nodes only
    const el = child as Element;
    if (el.tagName === 'property') {
      const name = el.getAttribute('name') ?? '';
      const value = el.getAttribute('value') ?? '';
      properties[name] = value;
    } else if (el.tagName === 'link') {
      const href = el.getAttribute('href') ?? '';
      functionLinks.push(href);
    }
  }

  // Find the root node element
  let rootNode: Node | null = null;
  for (let i = 0; i < root.childNodes.length; i++) {
    const child = root.childNodes[i];
    if (child.nodeType !== 1) continue;
    const el = child as Element;
    if (el.tagName === 'node') {
      rootNode = parseNodeElement(el, libraryLoader);
      break;
    }
  }

  if (!rootNode) {
    throw new Error('No root node found in ndbx');
  }

  return {
    formatVersion: CURRENT_FORMAT_VERSION,
    uuid,
    root: rootNode,
    properties,
    functionLinks,
  };
}

function parseNodeElement(el: Element, libraryLoader?: LibraryLoader): Node {
  const name = el.getAttribute('name') ?? '';
  const prototypeRef = el.getAttribute('prototype') ?? null;
  const func = el.getAttribute('function') ?? null;
  const category = el.getAttribute('category') ?? null;
  const description = el.getAttribute('description') ?? null;
  const image = el.getAttribute('image') ?? null;
  const posStr = el.getAttribute('position') ?? '0,0';
  const outputType = el.getAttribute('outputType') ?? null;
  const outputRange = (el.getAttribute('outputRange') as PortRange) ?? null;
  const renderedChild = el.getAttribute('renderedChild') ?? null;
  const handle = el.getAttribute('handle') ?? null;
  const comment = el.getAttribute('comment') ?? null;
  const alwaysRendered = el.getAttribute('alwaysRendered') === 'true';

  // Parse position
  const posParts = posStr.split(',');
  const position = {
    x: parseFloat(posParts[0]) || 0,
    y: parseFloat(posParts[1]) || 0,
  };

  // Resolve prototype
  let protoNode: Node | null = null;
  if (prototypeRef) {
    protoNode = resolvePrototype(prototypeRef, libraryLoader);
  }

  // Parse child elements
  const ports: Port[] = [];
  const children: Node[] = [];
  const connections: Connection[] = [];

  for (let i = 0; i < el.childNodes.length; i++) {
    const child = el.childNodes[i];
    if (child.nodeType !== 1) continue;
    const childEl = child as Element;

    if (childEl.tagName === 'port') {
      ports.push(parsePortElement(childEl));
    } else if (childEl.tagName === 'node') {
      children.push(parseNodeElement(childEl, libraryLoader));
    } else if (childEl.tagName === 'conn') {
      connections.push(parseConnectionElement(childEl));
    } else if (childEl.tagName === 'importCoreNode') {
      const ref = childEl.getAttribute('ref');
      if (ref === 'ROOT') {
        // ROOT is the base template — add no-op node with core/zero function
        children.push(createRootNode());
      } else if (ref === 'NETWORK') {
        children.push(createNetworkNode());
      }
    }
  }

  // Merge with prototype
  const node: Node = {
    name,
    prototype: prototypeRef,
    function: func ?? protoNode?.function ?? null,
    category: category ?? protoNode?.category ?? null,
    description: description ?? protoNode?.description ?? null,
    image: image ?? protoNode?.image ?? null,
    position,
    comment,
    inputs: mergeInputs(protoNode?.inputs ?? [], ports),
    outputType: outputType ?? protoNode?.outputType ?? 'geometry',
    outputRange: outputRange ?? protoNode?.outputRange ?? 'value',
    children,
    connections,
    renderedChild,
    handle: handle ?? protoNode?.handle ?? null,
    alwaysRendered,
  };

  return node;
}

function parsePortElement(el: Element): Port {
  const name = el.getAttribute('name') ?? '';
  const type = (el.getAttribute('type') ?? 'float') as PortType;
  const label = el.getAttribute('label') ?? null;
  const description = el.getAttribute('description') ?? null;
  const widget = (el.getAttribute('widget') ?? null) as Widget | null;
  const range = (el.getAttribute('range') ?? 'value') as PortRange;
  const valueStr = el.getAttribute('value') ?? null;
  const minStr = el.getAttribute('min') ?? null;
  const maxStr = el.getAttribute('max') ?? null;
  const childReference = el.getAttribute('childReference') ?? null;

  // Parse menu items
  const menuItems: MenuItem[] = [];
  for (let i = 0; i < el.childNodes.length; i++) {
    const child = el.childNodes[i];
    if (child.nodeType !== 1) continue;
    const menuEl = child as Element;
    if (menuEl.tagName === 'menu') {
      menuItems.push({
        key: menuEl.getAttribute('key') ?? '',
        label: menuEl.getAttribute('label') ?? '',
      });
    }
  }

  return {
    name,
    type,
    label,
    description,
    widget: widget ?? defaultWidgetForType(type),
    range,
    value: parsePortValue(valueStr, type),
    minimumValue: minStr !== null ? parseFloat(minStr) : null,
    maximumValue: maxStr !== null ? parseFloat(maxStr) : null,
    menuItems,
    childReference,
  };
}

function parseConnectionElement(el: Element): Connection {
  const input = el.getAttribute('input') ?? '';
  const output = el.getAttribute('output') ?? '';

  // input format: "nodeName.portName"
  const dotIdx = input.indexOf('.');
  const inputNode = dotIdx >= 0 ? input.substring(0, dotIdx) : input;
  const inputPort = dotIdx >= 0 ? input.substring(dotIdx + 1) : '';

  return {
    outputNode: output,
    inputNode,
    inputPort,
  };
}

function parsePortValue(str: string | null, type: PortType): Value {
  if (str === null || str === '') {
    return getDefaultValue(type);
  }
  switch (type) {
    case 'int': return { type: 'int', value: parseInt(str, 10) || 0 };
    case 'float': return { type: 'float', value: parseFloat(str) || 0 };
    case 'boolean': return { type: 'boolean', value: str === 'true' };
    case 'string': return { type: 'string', value: str };
    case 'point': {
      const parts = str.split(',');
      return {
        type: 'point',
        value: {
          x: parseFloat(parts[0]) || 0,
          y: parseFloat(parts[1]) || 0,
        },
      };
    }
    case 'color': {
      // Parse "#rrggbbaa" format
      const hex = str.startsWith('#') ? str.slice(1) : str;
      if (hex.length === 8) {
        return {
          type: 'color',
          value: {
            r: parseInt(hex.slice(0, 2), 16) / 255,
            g: parseInt(hex.slice(2, 4), 16) / 255,
            b: parseInt(hex.slice(4, 6), 16) / 255,
            a: parseInt(hex.slice(6, 8), 16) / 255,
          },
        };
      }
      if (hex.length === 6) {
        return {
          type: 'color',
          value: {
            r: parseInt(hex.slice(0, 2), 16) / 255,
            g: parseInt(hex.slice(2, 4), 16) / 255,
            b: parseInt(hex.slice(4, 6), 16) / 255,
            a: 1,
          },
        };
      }
      return { type: 'null' };
    }
    case 'geometry':
    case 'list':
    case 'data':
    case 'context':
      return { type: 'null' };
    default:
      return { type: 'string', value: str };
  }
}

function getDefaultValue(type: PortType): Value {
  switch (type) {
    case 'int': return { type: 'int', value: 0 };
    case 'float': return { type: 'float', value: 0 };
    case 'boolean': return { type: 'boolean', value: false };
    case 'string': return { type: 'string', value: '' };
    case 'point': return { type: 'point', value: { x: 0, y: 0 } };
    case 'color': return { type: 'color', value: { r: 0, g: 0, b: 0, a: 1 } };
    default: return { type: 'null' };
  }
}

function defaultWidgetForType(type: PortType): Widget {
  switch (type) {
    case 'int': return 'int';
    case 'float': return 'float';
    case 'string': return 'string';
    case 'boolean': return 'toggle';
    case 'point': return 'point';
    case 'color': return 'color';
    default: return 'none';
  }
}

function mergeInputs(protoInputs: Port[], overridePorts: Port[]): Port[] {
  // Start with proto's ports, then override by name
  const result: Port[] = protoInputs.map(p => ({ ...p }));

  for (const override of overridePorts) {
    const existing = result.findIndex(p => p.name === override.name);
    if (existing >= 0) {
      // Merge: override only the fields that are set
      result[existing] = {
        ...result[existing],
        ...override,
        // Keep proto values for fields not specified in override
        label: override.label ?? result[existing].label,
        description: override.description ?? result[existing].description,
      };
    } else {
      result.push(override);
    }
  }

  return result;
}

function resolvePrototype(ref: string, libraryLoader?: LibraryLoader): Node | null {
  const dotIdx = ref.indexOf('.');
  if (dotIdx < 0) return null;

  const libName = ref.substring(0, dotIdx);
  const nodeName = ref.substring(dotIdx + 1);

  // Try loading from cache
  let lib = libraryCache.get(libName);
  if (!lib && libraryLoader) {
    lib = libraryLoader(libName);
    if (lib) libraryCache.set(libName, lib);
  }

  if (!lib) return null;

  // Find node in library's root children
  return findNodeInChildren(lib.root, nodeName);
}

function findNodeInChildren(node: Node, name: string): Node | null {
  for (const child of node.children) {
    if (child.name === name) return child;
  }
  return null;
}

function createRootNode(): Node {
  return {
    name: 'ROOT',
    prototype: null,
    function: 'core/zero',
    category: null,
    description: 'Base node',
    image: null,
    position: { x: 0, y: 0 },
    comment: null,
    inputs: [],
    outputType: 'geometry',
    outputRange: 'value',
    children: [],
    connections: [],
    renderedChild: null,
    handle: null,
    alwaysRendered: false,
  };
}

function createNetworkNode(): Node {
  return {
    name: 'NETWORK',
    prototype: null,
    function: null,
    category: null,
    description: 'Network node',
    image: null,
    position: { x: 0, y: 0 },
    comment: null,
    inputs: [],
    outputType: 'geometry',
    outputRange: 'value',
    children: [],
    connections: [],
    renderedChild: null,
    handle: null,
    alwaysRendered: false,
  };
}

// Map function references from Java/Python to TypeScript
export function mapFunctionRef(ref: string): string {
  // Java: "corevector/rect" stays the same
  // Python: function="pyvector/compound" → "corevector/compound"
  if (ref.startsWith('pyvector/')) {
    return 'corevector/' + ref.substring('pyvector/'.length);
  }
  return ref;
}

function parseXml(xml: string): Document {
  if (typeof DOMParser !== 'undefined') {
    return new DOMParser().parseFromString(xml, 'text/xml');
  }
  // For Node.js, try @xmldom/xmldom
  try {
    const { DOMParser: XmlDOMParser } = require('@xmldom/xmldom');
    return new XmlDOMParser().parseFromString(xml, 'text/xml');
  } catch {
    throw new Error('XML parsing requires DOMParser (browser) or @xmldom/xmldom (Node.js)');
  }
}

// Allow clearing the cache for testing
export function clearLibraryCache(): void {
  libraryCache.clear();
}
